src/Leela.cpp

```diff
@@ -61,14 +61,82 @@ static void license_blurb() {
         PROGRAM_VERSION);
 }
 
+static void calculate_thread_count_cpu(boost::program_options::variables_map & vm) {
+    // If we are CPU-based, there is no point using more than the number of CPUs/
+    auto cfg_max_threads = std::min(SMP::get_num_cpus(), size_t{MAX_CPUS});
+
+    if (vm.count("threads")) {
+        auto num_threads = vm["threads"].as<unsigned int>();
+        if (num_threads > cfg_max_threads) {
+            myprintf("Clamping threads to maximum = %d\n", cfg_max_threads);
+            num_threads = cfg_max_threads;
+        }
+        cfg_num_threads = num_threads;
+    } else {
+        cfg_num_threads = cfg_max_threads;
+    }
+}
+
+#ifdef USE_OPENCL
+static void calculate_thread_count_gpu(boost::program_options::variables_map & vm) {
+    auto cfg_max_threads = size_t{MAX_CPUS};
+
+    // Default thread count : GPU case
+    // 1) if no args are given, use batch size of 5 and thread count of (batch size) * (number of gpus) * 2
+    // 2) if number of threads are given, use batch size of (thread count) / (number of gpus) / 2
+    // 3) if number of batches are given, use thread count of (batch size) * (number of gpus) * 2
+    auto gpu_count = cfg_gpus.size();
+    if (gpu_count == 0) {
+        // size of zero if autodetect GPU : default to 1
+        gpu_count = 1;
+    }
+
+    if (vm.count("threads")) {
+        auto num_threads = vm["threads"].as<unsigned int>();
+        if (num_threads > cfg_max_threads) {
+            myprintf("Clamping threads to maximum = %d\n", cfg_max_threads);
+            num_threads = cfg_max_threads;
+        }
+        cfg_num_threads = num_threads;
+
+        if (vm.count("batchsize")) {
+            cfg_batch_size = vm["batchsize"].as<unsigned int>();
+        } else {
+            cfg_batch_size = (cfg_num_threads + (gpu_count * 2) - 1) / (gpu_count * 2);
```

Comment:
Why the `(gpu_count * 2) - 1`?

I think `cfg_num_threads / (gpu_count + 1)` is reasonable, since it's theoretically possible to saturate *n* gpus with *(n+1)* x batchsize search threads (if the gpus are of the same speed with evaluations scheduled in an evenly spaced manner, and if the search threads are fast enough to pick up a position in *1/n* of the time of a batch evaluation).